src/coin/cgld/stakingBuilder.ts

```
@@ -58,62 +69,245 @@ export class StakingBuilder {
+        this.validateAmount();
+        return this.buildUnlockStaking();
+      case StakingOperationTypes.WITHDRAW:
+        this.validateIndex();
+        return this.buildWithdrawStaking();
       default:
         throw new InvalidTransactionError('Invalid staking operation: ' + this._type);
     }
   }
 
+  /**
+   * Builds a lock gold operation sending the amount on the transaction value field
+   *
+   * @returns {StakingCall} a lock gold operation using the LockedGold contract
+   */
+  private buildLockStaking(): StakingCall {
+    const operation = getOperationConfig(this._type, this._coinConfig.network.type);
+    return new StakingCall(this._amount, operation.contractAddress, operation.methodId, operation.types, []);
+  }
+
+  /**
+   * Builds an unlock gold operation sending the amount encoded on the data field
+   *
+   * params
+   * amount: amount of locked gold to be unlocked
+   *
+   * @returns {StakingCall} an unlock gold operation using the LockedGold contract
+   */
+  private buildUnlockStaking(): StakingCall {
+    const operation = getOperationConfig(this._type, this._coinConfig.network.type);
+    const params = [this._amount];
+    return new StakingCall('0', operation.contractAddress, operation.methodId, operation.types, params);
+  }
+
+  /**
+   * Builds a vote operation that uses locked gold to add pending votes for a validator group.
+   *
+   * params
+   * validatorGroup: group to vote for
+   * amount: amount of votes (locked gold) for the group
+   * lesser: validator group that has less votes than the validatorGroup
+   * greater: validator group that has more vots than the validatorGroup
+   *
+   * @returns {StakingCall} an vote operation using the Election contract
+   */
+  private buildVoteStaking(): StakingCall {
+    const operation = getOperationConfig(this._type, this._coinConfig.network.type);
+    const params = [this._validatorGroup, this._amount, this._lesser, this._greater];
```

one last thing I realize - can we enforce validation of greater actually being larger than lesser? if they pass these mixed up, we should throw an error, as opposed to swapping values (or letting celo's chain handle it)

---

One issue with that is that we don't have the sorted list on our side. The lesser and greater are just addresses with no information about the number of votes that each of them has.
We are currently depending on celo contractkit implementation to get these values. Using this method:
https://github.com/celo-org/celo-monorepo/blob/b631899aac58b2554892354833affc8df99919ec/packages/contractkit/src/wrappers/Election.ts#L423

---

ah this is ok then!